app/workers/set_tenant_id_worker.rb

```diff
@@ -22,19 +27,38 @@ def self.validate_params(*relations)
       raise "you must pass relations to fix" if relations.empty?
       raise "Only relations #{RELATIONS} are supported" unless (relations - RELATIONS).empty?
     end
+
     delegate :validate_params, :to => :class
 
     def perform(*relations)
       validate_params(*relations)
 
-      Account.tenants.select(:id).find_each do |provider|
-        SetTenantIdWorker.perform_later(provider, relations)
+      enqueue_tenant_relations(relations)
+      enqueue_account_relations(relations)
+    end
+
+    def enqueue_scope(scope, relations)
+      scope.select(:id).find_each do |object|
+        SetTenantIdWorker.perform_later(object, relations)
       end
     end
+
+    def enqueue_tenant_relations(relations)
+      tenant_relations = relations & TENANT_RELATIONS
+      enqueue_scope(Account.tenants, tenant_relations) if tenant_relations.present?
+    end
+
+    def enqueue_account_relations(relations)
+      account_relations = relations & ACCOUNT_RELATIONS
+      enqueue_scope(Account.not_master, account_relations) if account_relations.present?
```

Comment:
Hm, does it mean that the master's alerts will never be "fixed"? Probably, it's fine, because it's a special case, where the tenant is the same as the account.

But as I am not sure whether there will be any assumption that the `tenant_id` is always present (whether the account is master or not).

Reply:
For master objects, `tenant_id` is supposed to be `nil`.

Reply:
FYI `tenants` scope also excludes master.

Reply:
Yeah, I noticed the `tenants` scope doesn't include master.

> For master objects, tenant_id is supposed to be nil.

OK, if so, then it's fine!

Reply:
If you check all triggers exclude master https://github.com/3scale/porta/blob/ec82abc74cbd9980eb38096b4d077b308021d4d3/lib/system/database/definitions/mysql.rb